contracts/Loans.vy

```diff
@@ -684,6 +711,24 @@ def reserve(
     if self.walletWhitelistEnabled and not self.walletsWhitelisted[msg.sender]:
         raise "msg.sender is not whitelisted"
 
+    # signature validation
+    collaterals_data_hash: DynArray[bytes32, 100] = []
+    for c in _collaterals:
+        collaterals_data_hash.append(keccak256(_abi_encode(COLLATERAL_TYPE_HASH, c.contractAddress, c.tokenId, c.amount)))
+
+    data_hash: bytes32 = keccak256(_abi_encode(
+                RESERVE_TYPE_HASH,
+                _amount,
+                _interest,
+                _maturity,
+                keccak256(slice(_abi_encode(collaterals_data_hash), 32*2, 32*len(_collaterals))),
+                _deadline,
+                ))
+
+    sig_hash: bytes32 = keccak256(concat(convert("\x19\x01", Bytes[2]), _abi_encode(self.reserve_sig_domain_separator, data_hash)))
+    signer: address = ecrecover(sig_hash, _v, _r, _s)
```

Comment:
Can we put this in a `_recoverSigner` method to keep the `reserve` method simpler?

Reply:
Done